### ISSUE TITLE
invincible zombie patch

### ISSUE THREAD

**Issue by jaxjaxk** [Zet0rz/nzombies#241]
i got some data on the invincible zombie recently. it seems the zombies health becomes negative for some reason. because of this we can easily run a timer to remove any zombie that has negative health every 5 seconds and while this does not fix the problem it DOES keep us from having to deal with it so much. you can also just have it check when the zombie takes damage or whatever to make it more seamless.whatever the case try this 

`

```
timer.Create( "kill invalid zombies", 5, 0, function()
if SERVER then
local c = {}
for k,v in pairs(nzConfig.ValidEnemies) do
table.Merge( c, ents.FindByClass(k) )

end
```

for i,j in ipairs(c) do
if !j:Health():invalid() or j:Health() <=0 then
    --ulx.logString("zombie ".. i .. " is not valid, removing" , true )
    j:Remove()
end

end
end

end)`

this should remove any zombies that have either nil health or health below 0. i also threw in something to log in ulx ( commented out) so that you can still see if the problem still occurs without having it disrupt gameplay.

**Comment by jaxjaxk** [comment]
another method would be to work it into the zombies think function. i did not make a pull request as their are multiple ways to do it and i am betting you will do a better job.

**Comment by Zet0rz** [comment]
I think making it part of their pathing/think functions is more efficient. But I think the most efficient would be to do it in the damage hook. So zombies may become invincible, but then die from the next damage they take (making them not invincible at all).

**Comment by jaxjaxk** [comment]
@Zet0rz i think in the path finding function would be better. the reason is that the health of the zombie does not go down when it is bugged not matter how much you shoot it. this means that it is possible that the take damage hook might not even run. just food for thought. also maybe instead of remove you should have it do the whole death animation thing so it does not just disappear.

**Comment by jaxjaxk** [comment]
i believe you might have to add a simple timer around what i have put there. the code is not working for me right now so i think it has to do with the timer breaking because it runs before the gamemode is ready.

**Comment by lolleko** [comment]
I tried timer and think solutions a while ago, they won't work I kinda forgot the reason lol. But they won't work or at least they will cause bugs at other places, for example Zombies dying twice and stuff like that.

**Comment by Zet0rz** [comment]
Timers and constant checks are as inefficient as checking can get. So while it will work, the logical solution would be to simply add this to the EntityTakeDamage hook we use to give points:

``` Lua
if ent:Health() < 0 then ent:Kill() end
```

Now the question is if the EntityTakeDamage hook even gets called in these scenarios, but I do believe it does since they still take damage, they are just in negative health and never cross the 0-border to get killed. If it doesn't work, adding this line to the zombie base's Think function would be the second-best solution, it'd kill them instantly if they should go negative (and then their Think function no longer runs).

**Comment by jaxjaxk** [comment]
@Zet0rz i will do what you said here and add in a bit to log if it ever gets triggered. that way we know.

**Comment by Zet0rz** [comment]
I've done this change locally: http://i.imgur.com/lilPtEf.png
But since I can't reproduce invulnerable zombies, I can't test if it works. Give it a try. Here's the code for easier copy-paste:
`if zombie:Health() < 0 then zombie:Kill(dmginfo) end`
in file gamemode/enemies/sv_hooks.lua

**Comment by Zet0rz** [comment]
Okay this has been fixed on my end. Fixed by adding the line:

``` Lua
if nzConfig.ValidEnemies[zombie:GetClass()] and zombie:Health() < 0 then zombie:SetHealth(1) end
```

into EntityTakeDamage hook. It will set the health back up to 1, making the damage (which is applied after this hook) kill it off like it would normally. Invulnerable zombies can still appear, but they'd "reset" every damage they take making it a **very** low chance for it to happen repeatedly.

**Comment by Zet0rz** [comment]
Fixed in 1af80034dc69ff94c99ddd6c17ce85698ecb7f2b
